libcst/metadata/name_provider.py

```diff
@@ -0,0 +1,166 @@
+        else:
+            return any(qn == name for qn in qualified_names)
+
+
+class QualifiedNameVisitor(cst.CSTVisitor):
+    def __init__(self, provider: "QualifiedNameProvider") -> None:
+        self.provider: QualifiedNameProvider = provider
+
+    def on_visit(self, node: cst.CSTNode) -> bool:
+        scope = self.provider.get_metadata(ScopeProvider, node, None)
+        if scope:
+            self.provider.set_metadata(node, scope.get_qualified_names_for(node))
+        else:
+            self.provider.set_metadata(node, set())
+        super().on_visit(node)
+        return True
+
+
+DOT_PY: Pattern[str] = re.compile(r"(__init__)?\.py$")
+
+
+def _module_name(root_path: Path, path: str) -> Optional[str]:
+    try:
+        return (
+            DOT_PY.sub("", str(Path(path).relative_to(root_path)))
+            .replace("/", ".")
+            .rstrip(".")
+        )
+    except ValueError:
+        # relative_to failed, we can't guess the module name
+        return None
+
+
+class FullyQualifiedNameProvider(BatchableMetadataProvider[Collection[QualifiedName]]):
+    """
+    Provide fully qualified names for CST nodes. Like :class:`QualifiedNameProvider`,
+    but the provided :class:`QualifiedName`s have absolute identifier names instead of
+    local to the current module.
+
+    This provider is initialized with the current module's fully qualified name, and
+    can be used with :class:`~libcst.metadata.FullRepoManager`. Compared to :class:`QualifiedNameProvider`, it also resolves relative imports.
+    """
+
+    METADATA_DEPENDENCIES = (QualifiedNameProvider,)
+
+    @classmethod
+    def gen_cache(
+        cls, root_path: Path, paths: List[str], timeout: Optional[int] = None
+    ) -> Mapping[str, object]:
+        cache = {path: _module_name(root_path, path) for path in paths}
+        return cache
+
+    def __init__(self, cache: str) -> None:
+        super().__init__(cache)
+        self.module_name: str = cache
+
+    def visit_Module(self, node: cst.Module) -> bool:
+        visitor = FullyQualifiedNameVisitor(self, self.module_name)
+        node.visit(visitor)
+        self.set_metadata(
+            node,
+            {QualifiedName(name=self.module_name, source=QualifiedNameSource.LOCAL)},
+        )
+        return True
+
+
+class FullyQualifiedNameVisitor(cst.CSTVisitor):
+    @staticmethod
+    def _fully_qualify(module_name: str, qname: QualifiedName) -> QualifiedName:
+        if qname.source == QualifiedNameSource.BUILTIN:
+            # builtins are already fully qualified
+            return qname
+        name = qname.name
+        if qname.source == QualifiedNameSource.IMPORT and not name.startswith("."):
+            # non-relative imports are already fully qualified
+            return qname
+        if name.startswith("."):
+            name = name.lstrip(".")
+            parts_to_strip = len(qname.name) - len(name)
+            module_name = ".".join(module_name.split(".")[: -1 * parts_to_strip])
```

Comment:
This part is not easy to understand, can you extract as a function and add some unit tests?